services/wallet/reactor.go

```diff
@@ -0,0 +1,204 @@
+package wallet
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// HeaderReader interface for reading headers using block number or hash.
+type HeaderReader interface {
+	HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error)
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+}
+
+// NewReactor creates instance of the Reactor.
+func NewReactor(db *Database, feed *event.Feed, client *ethclient.Client, address common.Address, chain *big.Int) *Reactor {
+	reactor := &Reactor{
+		db:      db,
+		client:  client,
+		feed:    feed,
+		address: address,
+	}
+	reactor.erc20 = NewERC20TransfersDownloader(client, address)
+	reactor.eth = &ETHTransferDownloader{
+		client:  client,
+		address: address,
+		signer:  types.NewEIP155Signer(chain),
+	}
+	return reactor
+}
+
+// Reactor listens to new blocks and stores transfers into the database.
+type Reactor struct {
+	client  HeaderReader
+	db      *Database
+	feed    *event.Feed
+	address common.Address
+
+	eth   *ETHTransferDownloader
+	erc20 *ERC20TransfersDownloader
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// Start runs reactor loop in background.
+func (r *Reactor) Start() error {
+	if r.quit != nil {
+		return errors.New("already running")
+	}
+	r.quit = make(chan struct{})
+	r.wg.Add(1)
+	go func() {
+		log.Info("wallet reactor started", "address", r.address.String())
+		r.loop()
+		r.wg.Done()
+	}()
+	return nil
+}
+
+// Stop stops reactor loop and waits till it exits.
+func (r *Reactor) Stop() {
+	if r.quit == nil {
+		return
+	}
+	close(r.quit)
+	r.wg.Wait()
+	r.quit = nil
+}
+
+func (r *Reactor) loop() {
+	var (
+		// TODO(dshulyak) parametrize
+		// pow block on main chain is mined once per ~14 seconds
+		// but we for tests we are using clique chain with immediate block signer
+		ticker = time.NewTicker(1 * time.Second)
```

Comment:
Let's solve it now with `Reactor` param. We can default it using a map of networkID to tick time.

Reply:
makes sense. will do it that way